### PR TITLE
Update WZTo4Q sample to 5f scheme

### PR DESCRIPTION
**Pull request by jennetd**

### CODE REVIEW

**Comment by sihyunjeon** [comment]
Did you actually check if this is working? there is top quark interfering the process and did not work when b was massless, for me

**Comment by jennetd** [comment]
I successfully generated the gridpack. I will try generating some events from it now

**Comment by sihyunjeon** [comment]
that's weird... in any case even if it worked,

How do you deal with tzq overlap?
`add process p p > w+ z j [QCD] @1`

for example if j is b here and you get one more qcd vertex from NLO effect,
you effectively get `p p > w+ z b j`
which can be from `p p > t z q`
as in the diagram in [1]

[1] https://www.science20.com/tommaso_dorigo/cms_observes_associated_production_of_top_quarks_and_z_bosons-235551

**Comment by sihyunjeon** [comment]
can you also share your gridpack? curious how it worked in your case anyways

**Comment by jennetd** [comment]
I didn't take any measures to remove the tzq part, but I think you are right that I don't want that. @yihui-lai we should probably go back to the 4f scheme. 

You can find the gridpack on lxplus here: /afs/[cern.ch/user/j/jdickins/public/](http://cern.ch/user/j/jdickins/public/) WZTo4Q_5f_amcatnloFxFx_el8_amd64_gcc10_CMSSW_12_4_8_tarball.tar.xz

**Comment by yihui-lai** [comment]
> I didn't take any measures to remove the tzq part, but I think you are right that I don't want that. @yihui-lai we should probably go back to the 4f scheme.
> 
> You can find the gridpack on lxplus here: /afs/[cern.ch/user/j/jdickins/public/](http://cern.ch/user/j/jdickins/public/) WZTo4Q_5f_amcatnloFxFx_el8_amd64_gcc10_CMSSW_12_4_8_tarball.tar.xz

Yeah I agree, we should go back to 4f scheme. Probably it's better to define `define j > j b b~` in madspin card in stead of the proc card?

**Comment by Saptaparna** [comment]
Is it not recommended to veto those events in the process card with a $ t or something like that?

**Comment by sihyunjeon** [comment]
i agree to what yihui says, although I don't think there is more diagrams that's going to be generated by doing it in the proc_card I am not sure, so just to be in the safe side.

@jennetd i think it is ok. there was madgraph author's concern with this but i don't fully recall (@agrohsje we had a call with olivier 2 years ago about this or maybe it was different thing ... do you recall what it was?). or maybe it was something totally irrelevant. at least i don't see any issue right now.

@Saptaparna $$ loses gauge invariance so it should be avoided unless we are sure about the consequences

**Comment by menglu21** [comment]
Hi @sihyunjeon I'm not sure if there is overlap with tzq, cuz with current syntax, the top quark can only occur in the loop with Z boson, however in tzq process, there is always an additional b quark to allow the non-loop top quark

**Comment by sihyunjeon** [comment]
@menglu21 what is the "current syntax" you are referring to? it was previously 5fs for both p and j, now it's only for j. former or latter?

**Comment by menglu21** [comment]
I mean the 5f

**Comment by DickyChant** [comment]
Hi I think the issue is that:

1. For tZq process, it is essentially single top thus you need **additional W exchange** to have this flavor changing (i.e. you need a W to have t + b in the final state) thus you have additional EW vertex, resulting in **QED = 3**
2. Meanwhile for diboson it is QED = 2 since you have two vertices for boson creation

I mean at MG level they will not interfere at diagram level in this sense.

**Comment by DickyChant** [comment]
But, another point is that, if you want to have 5f, then in terms of final state, there is overlap and whether that interference is significant is another question...

**Comment by sihyunjeon** [comment]
@DickyChant ah makes sense.

though your modelling seems wrong @jennetd i checked your gridpack and you keep your 4fs definition in your run card ` 4 = maxjetflavor` this will make wzb and wzq contributions differently behave in jet merging. i am not sure if this was the reason why your gridpack was successful (although it has a flaw in the modelling), but at least it failed for me.

**Review by menglu21 — COMMENTED** [review]
besides the comments inline, please move these 5f cards to directory 2017/13TeV/...

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/VV/WZTo4Q_4f_amcatnloFxFx/WZTo4Q_4f_amcatnloFxFx_proc_card.dat [hunk 1]
```diff
@@ -1,9 +1,6 @@
 import model loop_sm
 
-define ell+ = e+ mu+ ta+
-define ell- = e- mu- ta-
-define vl = ve vm vt
-define vl~ = ve~ vm~ vt~
+define j = j b b~
```
[the remark this answers is not in the document]
Comment (jennetd) [reply bin/MadGraph5_aMCatNLO/cards/production/13TeV/VV/WZTo4Q_4f_amcatnloFxFx/WZTo4Q_4f_amcatnloFxFx_proc_card.dat:3]: It's not really important to me whether b is included in the PDF, but I do want to have the Z > bb decay included in this sample. Is this not the recommended way to do this?

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/VV/WZTo4Q_5f_amcatnloFxFx/WZTo4Q_5f_amcatnloFxFx_run_card.dat [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  -1  = etagamma  ! Max photon abs(pseudo-rap)
+ 0.4  = R0gamma   ! Radius of isolation code
+ 1.0  = xn        ! n parameter of eq.(3.4) in hep-ph/9801442
+ 1.0  = epsgamma  ! epsilon_gamma parameter of eq.(3.4) in hep-ph/9801442
+ .true.  = isoEM  ! isolate photons from EM energy (photons and leptons)
+#***********************************************************************
+# Maximal PDG code for quark to be considered a jet when applying cuts.*
+# At least all massless quarks of the model should be included here.   *
+#***********************************************************************
+ 4 = maxjetflavor
```
[the remark this answers is not in the document]
Comment (jennetd) [reply bin/MadGraph5_aMCatNLO/cards/production/13TeV/VV/WZTo4Q_5f_amcatnloFxFx/WZTo4Q_5f_amcatnloFxFx_run_card.dat:144]: Just double-checking. You would prefer the 5f cards be here? https://github.com/jennetd/genproductions/tree/master/bin/MadGraph5_aMCatNLO/cards/production/2017/13TeV

**Comment by sihyunjeon** [comment]
If you want 5FS (b included in p), you need `maxjetflavor=5` (hence your gridpack you had previously is wrongly set and i presume this was why your gridpack was wrongly successful which should've failed)
If you want 4FS but still want `z > bb`, keep `maxjetflavor=4` but do `define q = j b b~` and do `decay z > q q`

**Comment by jennetd** [comment]
Ok, in this case I will remove the 5f cards entirely since the gridpack will fail, and concentrate on the 4f ones. I think these should be correct.

**Comment by sihyunjeon** [comment]
> Ok, in this case I will remove the 5f cards entirely since the gridpack will fail, and concentrate on the 4f ones. I think these should be correct.

for safety, which @yihui-lai mentioned, can you at least check how many feynman diagrams are collected by doing
`j = j b b~` and not doing it WITHOUT using madspin?
i am not sure if there is, BUT if there is a diagram chiming in by doing `j = j b b~` you are in the middle of 4 and 5FS (proton definition is 4FS but from the jet merging perspective it is 5FS)

Otherwise, just remove `j = j b b~` in the proc_card and instead add it into the madspin so that nothing happens in the proc_card level

**Comment by jennetd** [comment]
Sure, I will try this

**Comment by jennetd** [comment]
Hi again, I tested this. Without using madspin, I see the same cross section and same number of diagrams regardless of whether I include j > j b

**Comment by DickyChant** [comment]
I am a little bit puzzled by this merge request, what is the goal of "moving" to 5f? 

1. if you want to have Z decays to bb~, you can do it with solely with madspin by redefine `multiparticles` there (saying `j = j b b~`_
2. if you want also the additional parton to also have b quark I dont think it is justified to have `j = j b b~`, even if you see the same numbers of diagram which comes from the fact that you have W and ckm is diagonal here. Then you see there is no b quark coming into play, why not keep everything in a 4f manner?

**Comment by jennetd** [comment]
The title of the MR is out of date - we decided to keep the 4f so that there is no b quark in the proton. The question was then is j = j b b~ the right way to add b quark to the Z decay. It sounds like you @DickyChant are suggesting I do this in madspin rather than the proc card. That would be fine for me, I just wasn't sure what is recommended.

**Comment by DickyChant** [comment]
> The title of the MR is out of date - we decided to keep the 4f so that there is no b quark in the proton. The question was then is j = j b b~ the right way to add b quark to the Z decay. It sounds like you @DickyChant are suggesting I do this in madspin rather than the proc card. That would be fine for me, I just wasn't sure what is recommended.

I believe a benefit that we have this git repository is to make sure every piece of merge request would serve as an example for other cms colleagues.

Doing `j = j b b~` without changing `p` is for sure not a good practice and should not be advertised in general. MadGraph would sometimes overwritten the two definitions simultaneously to keep things consistent. 

Since there is no benefit of using `j = j b b~` here, I see no point of doing it in proc_card. 

In fact you could put sth in your madspin card like ` jb = j b b~` followed by ` decay Z > jb jb`, here `jb` could be whatever code name you prefer. 

In case you don't believe, I just double checked with simple pp > z with z decay to quarks and it worked as [here](https://sqian.web.cern.ch/sqian/test_znlo/)

**Comment by jennetd** [comment]
I am happy with this proposal. I will adjust the cards and generate the gridpack.

**Comment by DickyChant** [comment]
Thanks.

For your process those discussed definitions, though look different, might lead to the very same thing. But I just think that we should keep the minimum ambiguity such that we don't get confused at some moment.

For example, in CMS gridpack generation workflow we have this line https://github.com/cms-sw/genproductions/blob/78953df4692afed3467dbbccd4e53601e62fcf52/bin/MadGraph5_aMCatNLO/gridpack_generation.sh#L342

which detects whether the process is using 5 flavor scheme. Then what if we want to impose more detailed check at some time to further check (e.g. j definition, maxjetflavor setup etc.). Maybe I am just overly worried, but I think keep things as neat as possible would help eventually.

**Comment by jennetd** [comment]
Hi again, I am looking at the analogous cards for WZToLNu2Q: https://github.com/cms-sw/genproductions/tree/master/bin/MadGraph5_aMCatNLO/cards/production/13TeV/VV/WZToLNu2Q_4f_amcatnloFxFx
It looks like the Z(bb) decay is not included there. The WZToLNu2Q sample has already been generated. 

In addition to the above changes for WZTo4Q, I would propose to add a dedicated WZToLNu2B sample to supplement the existing one.

**Comment by DickyChant** [comment]
> Hi again, I am looking at the analogous cards for WZToLNu2Q: https://github.com/cms-sw/genproductions/tree/master/bin/MadGraph5_aMCatNLO/cards/production/13TeV/VV/WZToLNu2Q_4f_amcatnloFxFx It looks like the Z(bb) decay is not included there. The WZToLNu2Q sample has already been generated.
> 
> In addition to the above changes for WZTo4Q, I would propose to add a dedicated WZToLNu2B sample to supplement the existing one. Should this be a separate MR?

It is not my (MG contact's) business but at least I can speak for EXO sample requests would get started by analysts (saying preparing gridpacks and fragments), should this be propagated to the corresponding analyzers since we seem to talk about run2?

**Comment by jennetd** [comment]
I am an analyst on all hadronic VH H(bb), so I do need these samples to include Z(bb) decay. That is why I am working on the samples - was just saying I need to add Z(bb) to another channel as well

**Comment by DickyChant** [comment]
> I am an analyst on all hadronic VH H(bb), so I do need these samples to include Z(bb) decay. That is why I am working on the samples - was just saying I need to add Z(bb) to another channel as well

Perfect thanks. 

Speaking of PR I dont think we have a strict rule of 1 PR per sample process ;)

**Comment by jennetd** [comment]
Ok! I have made the suggested changes to WZTo4Q and added WZToLNu2B.  I have gridpacks on lxplus for @yihui-lai when you are ready.

**Comment by yihui-lai** [comment]
Hi @jennetd, where are the new gridpacks? I can already take a look before the cards get approved.

**Comment by jennetd** [comment]
Hi @yihui-lai, they are here on lxplus: /afs/cern.ch/user/j/jdickins/public/

The names are
WZTo4Q_4f_amcatnloFxFx_el8_amd64_gcc10_CMSSW_12_4_8_tarball.tar.xz
WZToLNu2B_4f_amcatnloFxFx_el8_amd64_gcc10_CMSSW_12_4_8_tarball.tar.xz

**Comment by yihui-lai** [comment]
> @DickyChant @sihyunjeon @menglu21 Is there anything else I need to do before this is merged? Thanks.

Hi @menglu21, we (Hbb MC contact) checked the gridpacks as well and they look good. If there's no other concerns, could you merge the PR and we can start sample requests? Thanks!

**Pull request merged**